### PR TITLE
Add new --publish option

### PR DESCRIPTION
**Pull request by aslakhellesoy**
This adds a new `--publish` option. It does the same as `--format message:URL`, but it is simpler for users to use correctly.

The URL where the user can access the published report is printed to STDOUT.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,11 +14,17 @@ Please visit [cucumber/CONTRIBUTING.md](https://github.com/cucumber/cucumber/blo
 
 ### Added
 
-* N/A
+* `--publish` automatically publishes reports to [reports.cucumber.io](https://reports.cucumber.io)
 
 ### Changed
 
-* N/A
+* Updated gems:
```
Comment (aslakhellesoy) [inline CHANGELOG.md:21]: Update these again

File: lib/cucumber/cli/options.rb [hunk 2]
```diff
@@ -9,6 +9,7 @@
 module Cucumber
   module Cli
     class Options
+      CUCUMBER_MESSAGE_STORE_URL = ENV['CUCUMBER_MESSAGE_STORE_URL'] || 'https://messages.cucumber.io'
```
[the remark this answers is not in the document]
Comment (cbliard) [reply lib/cucumber/cli/options.rb]: done

File: lib/cucumber/cli/options.rb [hunk 3]
```diff
@@ -93,6 +94,7 @@ def parse!(args) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
 
         @args.options do |opts| # rubocop:disable Metrics/BlockLength
           opts.banner = banner
+          opts.on('--publish', 'Publish a report to https://reports.cucumber.io') { @options[:formats] << ['message', {}, "#{CUCUMBER_MESSAGE_STORE_URL}/api/reports"] }
```
[the remark this answers is not in the document]
Comment (cbliard) [reply lib/cucumber/cli/options.rb]: change + append path is done

adding auth header not done yet

File: spec/cucumber/formatter/http_io_spec.rb [hunk 4]
```diff
@@ -80,16 +80,16 @@ module Formatter
       context 'created by Io#ensure_io' do
         it 'returns a IOHTTPBuffer' do
           url = start_server
-          io = ensure_io("#{url}/ -X POST")
+          io = ensure_io("#{url}/s3 -X POST")
```
[the remark this answers is not in the document]
Comment (cbliard) [reply spec/cucumber/formatter/http_io_spec.rb:83]: that one is not that obvious because WebRick does not support `PUT` by default. I won't be able to finish it for today https://stackoverflow.com/questions/4996170/handle-the-put-method-in-webrick

File: features/docs/cli/publish_banner.feature [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    Feature: Hello
+    """
+    When I run `cucumber`
+    Then the output should contain:
+    """
+    ┌──────────────────────────────────────────────────────────────────────────┐
+    │ Share your Cucumber Report with your team at https://reports.cucumber.io │
+    │                                                                          │
+    │ Command line option:    --publish                                        │
+    │ Environment variable:   CUCUMBER_PLUGIN_PUBLISH_ENABLED=true             │
```
Comment (aslakhellesoy) [inline features/docs/cli/publish_banner.feature:18]: Change to `CUCUMBER_PUBLISH_ENABLED` when https://github.com/cucumber/cucumber-jvm/pull/2089 is merged

File: features/docs/cli/publish_banner.feature [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    │                                                                          │
```
Comment (aslakhellesoy) [inline features/docs/cli/publish_banner.feature:19]: Add a 3rd option

```
cucumber.yml:      default: --publish
```

I know it's similar to option 1, but it might help with adoption.

File: lib/cucumber/formatter/publish_banner_printer.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+              ],
+              '',
+              [
+                'More information at ',
+                ['https://reports.cucumber.io/docs/cucumber-ruby', :cyan]
+              ],
+              '',
+              [
+                'To disable this message, specify ',
+                ['CUCUMBER_PUBLISH_QUIET=true', :dark],
```
Comment (aslakhellesoy) [inline lib/cucumber/formatter/publish_banner_printer.rb:35]: Change to `:bold` because `:dark` (grey??) isn't legible in black terminals. Fixed in Cucumber-JVM: https://github.com/cucumber/cucumber-jvm/commit/52c85472c78a52ac7174460f181e3f2e7f990b90

File: lib/cucumber/formatter/url_reporter.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      def initialize(io)
+        @io = io
+      end
+
+      def report(url)
+        uri = URI(url)
+        display_banner(
+          [
+            'View your Cucumber Report at:',
+            [["https://reports.cucumber.io#{uri.path}", :blue, :underline]]
```
Comment (aslakhellesoy) [inline lib/cucumber/formatter/url_reporter.rb:17]: Cucumber-JVM's `NoPublishFormatter` uses `:cyan` + `:bold` (and no underline). Some terminals (iTerm2) will underline URLs automatically, but maybe we should use underline in Cucumber-JVM too?

So agree on `:cyan` + `:bold` + `:underline`?

Comment (aslakhellesoy) [reply lib/cucumber/formatter/url_reporter.rb:17]: I've updated Cucumber-JVM to do `:cyan` + `:bold` + `:underline` now

File: spec/cucumber/formatter/http_io_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (aslakhellesoy) [reply spec/cucumber/formatter/http_io_spec.rb:83]: Try adding the `webrick` gem, it might be more recent than the one supplied in the stdlib, and might support it.

File: lib/cucumber/cli/options.rb [hunk 9]
```diff
@@ -93,6 +94,8 @@ def parse!(args) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
 
         @args.options do |opts| # rubocop:disable Metrics/BlockLength
           opts.banner = banner
+          opts.on('--publish', 'Publish a report to https://reports.cucumber.io') { @options[:formats] << ['message', {}, CUCUMBER_PLUGIN_PUBLISH_URL] }
```
Comment (aslakhellesoy) [inline lib/cucumber/cli/options.rb:97]: Add authentication header here if `CUCUMBER_PUBLISH_TOKEN` is present. Like this:

https://github.com/cucumber/cucumber-jvm/blob/1ba0b255bdf3d4206146d6834096f69a446ffc4c/core/src/main/java/io/cucumber/core/plugin/PublishFormatter.java#L56

File: lib/cucumber/cli/options.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (aslakhellesoy) [reply lib/cucumber/cli/options.rb]: @mpkorstanje seems to agree with env var renames in https://github.com/cucumber/cucumber-jvm/pull/2089 so let's go with this.

File: spec/cucumber/formatter/http_io_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbliard) [reply spec/cucumber/formatter/http_io_spec.rb]: webrick gem was already used. I have aliased `do_PUT` to `do_GET` in webrick internal classes to get it working.

File: lib/cucumber/cli/options.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (cbliard) [reply lib/cucumber/cli/options.rb]: rename done

adding auth header still to be done

File: features/docs/cli/publish_banner.feature (hunk 5 above)
Comment (cbliard) [reply features/docs/cli/publish_banner.feature]: done

File: features/docs/cli/publish_banner.feature (hunk 6 above)
Comment (cbliard) [reply features/docs/cli/publish_banner.feature:20]: done

File: lib/cucumber/formatter/publish_banner_printer.rb (hunk 7 above)
Comment (cbliard) [reply lib/cucumber/formatter/publish_banner_printer.rb]: done

File: lib/cucumber/formatter/url_reporter.rb (hunk 8 above)
Comment (cbliard) [reply lib/cucumber/formatter/url_reporter.rb:17]: done too

File: lib/cucumber/cli/options.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (cbliard) [reply lib/cucumber/cli/options.rb]: done

File: lib/cucumber/cli/options.rb (hunk 9 above)
Comment (cbliard) [reply lib/cucumber/cli/options.rb]: done

File: cucumber.gemspec [hunk 10]
```diff
@@ -22,10 +22,10 @@ Gem::Specification.new do |s|
   s.add_dependency 'builder', '~> 3.2', '>= 3.2.3'
   s.add_dependency 'cucumber-core', '~> 7.1', '>= 7.1.0'
   s.add_dependency 'cucumber-create-meta', '~> 1.0.0', '>= 1.0.0'
-  s.add_dependency 'cucumber-cucumber-expressions', '~> 10.1', '>= 10.1.0'
-  s.add_dependency 'cucumber-gherkin', '~> 14.0', '>= 14.0.1'
-  s.add_dependency 'cucumber-html-formatter', '~> 7.0', '>= 7.0.0'
-  s.add_dependency 'cucumber-messages', '~> 12.2', '>= 12.2.0'
+  s.add_dependency 'cucumber-cucumber-expressions', '~> 10.2', '>= 10.2.2'
```
[the remark this answers is not in the document]
Comment (cbliard) [reply cucumber.gemspec]: done

File: lib/cucumber/formatter/url_reporter.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    class URLReporter
+      include Term::Banner
+
+      def initialize(io)
+        @io = io
+      end
+
+      def report(url)
+        uri = URI(url)
+        display_banner(
```
[the remark this answers is not in the document]
Comment (cbliard) [reply lib/cucumber/formatter/url_reporter.rb:14]: done

File: lib/cucumber/formatter/io.rb [hunk 12]
```diff
@@ -11,7 +13,8 @@ def ensure_io(path_or_url_or_io)
         return nil if path_or_url_or_io.nil?
         return path_or_url_or_io if path_or_url_or_io.respond_to?(:write)
         io = if path_or_url_or_io.match(%r{^https?://})
-               HTTPIO.open(path_or_url_or_io)
+               reporter = path_or_url_or_io.start_with?(Cucumber::Cli::Options::CUCUMBER_MESSAGE_STORE_URL) ? URLReporter.new($stdout) : nil
```
[the remark this answers is not in the document]
Comment (cbliard) [reply lib/cucumber/formatter/io.rb]: I added a `NoReporter` object but I'm not sure that's what you had in mind https://github.com/cucumber/cucumber-ruby/pull/1452/commits/2630502f8343890c397e967ecd0017da2087bc52

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -59,6 +59,7 @@ help us to correct style violations reported here:
 
 ## Release Process
 
+* Upgrade gems with `scripts/update-gemspec`
```
Comment (luke-hill) [inline CONTRIBUTING.md:62]: I'm not sure I agree with this being done in most / all PR's. Isolating gem updates feels best done to separate PR's.

File: lib/cucumber/formatter/io.rb [hunk 14]
```diff
@@ -9,11 +11,15 @@ module Io
...
-               HTTPIO.open(path_or_url_or_io)
+        return path_or_url_or_io if io?(path_or_url_or_io)
+
+        io = if url?(path_or_url_or_io)
+               url = path_or_url_or_io
+               reporter = url.start_with?(Cucumber::Cli::Options::CUCUMBER_PUBLISH_URL) ? URLReporter.new($stdout) : NoReporter.new
+               HTTPIO.open(url, nil, reporter)
              else
-               File.open(path_or_url_or_io, Cucumber.file_mode('w'))
+               path = path_or_url_or_io
```
Comment (luke-hill) [inline lib/cucumber/formatter/io.rb:21]: can remove this local assign to reduce complexity down

File: lib/cucumber/runtime.rb [hunk 15]
```diff
@@ -166,11 +166,13 @@ def set_encoding
 
     require 'cucumber/formatter/ignore_missing_messages'
     require 'cucumber/formatter/fail_fast'
+    require 'cucumber/formatter/publish_banner_printer'
     require 'cucumber/core/report/summary'
     def report
```
Comment (luke-hill) [inline lib/cucumber/runtime.rb:171]: could we pop a newline above this to separate require concerns

File: lib/cucumber/term/banner.rb [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        end
+
+        def apply_modifiers(str, modifiers)
+          display = str
+          modifiers.each { |modifier| display = send(modifier, display) }
+          display
+        end
+
+        def line_length(line)
+          if line.is_a?(Array)
```
Comment (luke-hill) [inline lib/cucumber/term/banner.rb:43]: if you flip this conditional you can lose some lines.

```
return line.length unless line.is_a?(Array)

# other logic without if/end
```

File: CONTRIBUTING.md (hunk 13 above)
Comment (cbliard) [reply CONTRIBUTING.md:62]: Indeed. Maybe we'll do it in a separate PR next time. We'll discuss that with @aslakhellesoy.

**Comment by cbliard** [comment]
Thanks for your review @luke-hill. We just merged the branch so we integrated your review on master in 7c787ff1.

File: lib/cucumber/term/banner.rb (hunk 16 above)
Comment (cbliard) [reply lib/cucumber/term/banner.rb:43]: I prefer staying away from unless when possible because I find it harder to read. I also like having `else` when both conditions branch are equivalent (when there are no guard clauses).

So I rewrote it like this:

```ruby
if line.is_a?(Array)
  line.map { |span| span_length(span) }.sum
else
  line.length
end
```

File: lib/cucumber/formatter/io.rb (hunk 14 above)
Comment (cbliard) [reply lib/cucumber/formatter/io.rb:21]: done on master

File: lib/cucumber/runtime.rb (hunk 15 above)
Comment (cbliard) [reply lib/cucumber/runtime.rb:171]: done on master

File: lib/cucumber/term/banner.rb (hunk 16 above)
Comment (luke-hill) [reply lib/cucumber/term/banner.rb:43]: That's fine, but the code is a little inconsistent, so I tried to pick the one you went for most. I'm happy with either way I guess, but you've already merged it.

Comment (cbliard) [reply lib/cucumber/term/banner.rb:43]: Yes, it is a little inconsistent. Each developer code style evolves over time. Rubocop is here to help us with it thanksfully.
